pkg/logs/input/container/kubernetes_scanner.go

```diff
@@ -0,0 +1,181 @@
+	podExpiration = 20 * time.Second
+)
+
+// KubeScanner looks for new and deleted pods to start or stop one file tailer per container.
+type KubeScanner struct {
+	watcher            *kubelet.PodWatcher
+	sources            *config.LogSources
+	sourcesByContainer map[string]*config.LogSource
+	stopped            chan struct{}
+}
+
+// NewKubeScanner returns a new scanner.
+func NewKubeScanner(sources *config.LogSources, pp pipeline.Provider, auditor *auditor.Auditor) (*KubeScanner, error) {
+	// initialize a pod watcher to list new and expired pods
+	watcher, err := kubelet.NewPodWatcher(podExpiration)
+	if err != nil {
+		return nil, err
+	}
+	// initialize the tagger to collect container tags
+	err = tagger.Init()
+	if err != nil {
+		return nil, err
+	}
+	// initialize a file scanner to collect logs from container files.
+	return &KubeScanner{
+		watcher:            watcher,
+		sources:            sources,
+		sourcesByContainer: make(map[string]*config.LogSource),
+		stopped:            make(chan struct{}),
+	}, nil
+}
+
+// Start starts the scanner
+func (s *KubeScanner) Start() {
+	log.Info("Starting Kubernetes scanner")
+	go s.run()
+}
+
+// Stop stops the scanner
+func (s *KubeScanner) Stop() {
+	log.Info("Stopping Kubernetes scanner")
+	s.stopped <- struct{}{}
+}
+
+// run runs periodically a scan to detect new and deleted pod.
+func (s *KubeScanner) run() {
+	ticker := time.NewTicker(podScanPeriod)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			s.scan()
+		case <-s.stopped:
+			return
+		}
+	}
+}
+
+// scan handles new and deleted pods.
+func (s *KubeScanner) scan() {
+	s.updatePods()
+	s.expirePods()
+}
+
+// updatePods pulls the new pods and creates new log sources.
+func (s *KubeScanner) updatePods() {
+	pods, err := s.watcher.PullChanges()
+	if err != nil {
+		log.Error("can't list changed pods", err)
+		return
+	}
+	for _, pod := range pods {
+		if pod.Status.Phase == "Running" {
+			log.Infof("adding pod: %v", pod.Metadata.Name)
+			s.addSources(pod)
+		}
+	}
+}
+
+// expirePods fetches all expired pods and removes the corresponding log sources.
+func (s *KubeScanner) expirePods() {
+	entityIDs, err := s.watcher.Expire()
+	if err != nil {
+		log.Error("can't list expired pods", err)
+		return
+	}
+	for _, entityID := range entityIDs {
+		log.Infof("removing pod %v", entityID)
+		pod, err := s.watcher.GetPodForEntityID(entityID)
+		if err != nil {
+			log.Errorf("can't find pod %v: %v", entityID, err)
+			continue
+		}
+		s.removeSources(pod)
+	}
+}
+
+// addSources creates a new log source for each container of a new pod.
+func (s *KubeScanner) addSources(pod *kubelet.Pod) {
+	for _, container := range pod.Status.Containers {
+		containerID := container.ID
+		if _, exists := s.sourcesByContainer[containerID]; exists {
+			continue
+		}
+		source := s.getSource(pod, container)
+		s.sourcesByContainer[containerID] = source
+		s.sources.AddSource(source)
```

Comment:
don't we need to start tailing the logs?

Reply:
Oh, I get it, we add a file source to the list of sources so the file scanner picks up the tailing. Definitely worth a comment here, and an explanation at the beginning of the class